### PR TITLE
Improve lobbies connections

### PR DESCRIPTION
**Pull request by ClementPasteau**
Main things are:
- A player is now considered connected only after a couple heartbeats, allowing to start the lobby faster (or slower) and not based on a random countdown
- A player can now join a lobby mid-game, if defined in the lobby configuration in the game dashboard.
  - New actions & conditions help in case there is a need to handle a new player coming (though the default events should be enough to handle a new player)

### CODE REVIEW

File: Extensions/Multiplayer/messageManager.ts [hunk 1]
```diff
@@ -98,7 +98,8 @@ namespace gdjs {
     let lastSentGameSyncData: GameNetworkSyncData | null = null;
     let numberOfForcedGameUpdates = 0;
 
-    // Send heartbeat messages to host to ensure the connection is still alive.
+    // Send heartbeat messages from host to players, sensuring their connection is still alive,
```
Comment (4ian) [inline Extensions/Multiplayer/messageManager.ts:101]:
```suggestion
    // Send heartbeat messages from host to players, ensuring their connection is still alive,
```

**Review by 4ian — APPROVED** [review]

File: Extensions/Multiplayer/JsExtension.js [hunk 2]
```diff
@@ -443,7 +443,108 @@ module.exports = {
       .addIncludeFile('Extensions/Multiplayer/messageManager.js')
       .addIncludeFile('Extensions/Multiplayer/multiplayerVariablesManager.js')
       .addIncludeFile('Extensions/Multiplayer/multiplayertools.js')
-      .setFunctionName('gdjs.multiplayerMessageManager.hasPlayerLeft');
+      .setFunctionName('gdjs.multiplayerMessageManager.hasPlayerJustLeft');
+
+    extension
+      .addExpression(
+        'LastPlayerLeftNumber',
```
Comment (4ian) [inline Extensions/Multiplayer/JsExtension.js:450]: `LastLeavingPlayerNumber`?

File: Extensions/Multiplayer/multiplayertools.ts [hunk 3]
```diff
@@ -997,19 +961,23 @@ namespace gdjs {
             throw new Error('Malformed message.');
           }
 
-          handleLobbyJoinEvent(runtimeScene, event.data.lobbyId);
+          handleJoinLobbyEvent(runtimeScene, event.data.lobbyId);
```
Comment (4ian) [inline Extensions/Multiplayer/multiplayertools.ts:964]: Nice renamings

File: Extensions/Multiplayer/multiplayertools.ts [hunk 4]
```diff
@@ -238,28 +209,33 @@ namespace gdjs {
...
+    const handleJoiningPlayer = (runtimeScene: gdjs.RuntimeScene) => {
+      const lastestPlayerWhoJustJoined = gdjs.multiplayerMessageManager.getLatestPlayerWhoJustJoined();
+      if (lastestPlayerWhoJustJoined) {
+        const playerUsername = getPlayerUsername(lastestPlayerWhoJustJoined);
+        gdjs.multiplayerComponents.displayPlayerJoinedNotification(
+          runtimeScene,
+          playerUsername
+        );
       }
+      // We remove the players who just joined 1 by 1, so that they can be treated in different frames.
```
Comment (4ian) [inline Extensions/Multiplayer/multiplayertools.ts:235]: Great comment

Comment (4ian) [reply Extensions/Multiplayer/multiplayertools.ts:235]: (though here it's "joined" ;))

File: Extensions/Multiplayer/multiplayertools.ts [hunk 5]
```diff
@@ -166,17 +171,9 @@ namespace gdjs {
...
-        return _lobby.players.length;
-      }
-
-      // If the game has started, look at the pings received from the players.
-      if (_isLobbyGameRunning) {
-        return gdjs.multiplayerMessageManager.getNumberOfConnectedPlayers();
-      }
-
-      return 0;
+      // Wether the lobby game has started or not, the number of players in the lobby
```
Comment (4ian) [inline Extensions/Multiplayer/multiplayertools.ts:174]:
```suggestion
      // Whether the lobby game has started or not, the number of players in the lobby
```

File: Extensions/Multiplayer/multiplayertools.ts [hunk 6]
```diff
@@ -66,15 +65,18 @@ namespace gdjs {
...
     );
 
     gdjs.registerRuntimeScenePostEventsCallback(
       (runtimeScene: gdjs.RuntimeScene) => {
         if (disableMultiplayerForTesting) return;
 
+        handleLeavingPlayer(runtimeScene);
+        handleJoiningPlayer(runtimeScene);
+        // Joining players are handled in the heartbeats directly.
+        gdjs.multiplayerMessageManager.handleHeartbeatsReceived();
```
Comment (4ian) [inline Extensions/Multiplayer/multiplayertools.ts:78]: I don't understand this comment telling that we handled joining players in the heartbeats, while we just before have a function being called that is named "handleJoiningPlayer".

File: Extensions/Multiplayer/JsExtension.js [hunk 7]
```diff
@@ -443,7 +443,108 @@ module.exports = {
...
+      )
+      .addIncludeFile('Extensions/Multiplayer/multiplayercomponents.js')
+      .addIncludeFile('Extensions/Multiplayer/messageManager.js')
+      .addIncludeFile('Extensions/Multiplayer/multiplayerVariablesManager.js')
+      .addIncludeFile('Extensions/Multiplayer/multiplayertools.js')
+      .setFunctionName('gdjs.multiplayerMessageManager.hasPlayerJustJoined');
+
+    extension
+      .addExpression(
+        'LastPlayerJoinedNumber',
```
Comment (4ian) [inline Extensions/Multiplayer/JsExtension.js:526]: LastJoinedPlayerNumber?

File: Extensions/Multiplayer/JsExtension.js (hunk 2 above)
Comment (4ian) [reply Extensions/Multiplayer/JsExtension.js:450]: Or `LastLeftPlayerNumber`

File: Extensions/Multiplayer/messageManager.ts [hunk 8]
```diff
@@ -1905,19 +2082,26 @@ namespace gdjs {
...
 
     const getPlayerPing = (playerNumber: number) => {
-      if (playerNumber < 1) {
-        // Player 1 is the host, so we don't need to compute the ping.
-        // Any negative number is invalid.
-        return 0;
-      }
-
-      return _playersPings[playerNumber] || 0;
+      return (_playersInfo[playerNumber] || {}).ping || 0;
```
Comment (4ian) [inline Extensions/Multiplayer/messageManager.ts:2104]: Will this crash if playerNumber is out of bound? Should we protect this or consider it's ok?

File: Extensions/Multiplayer/multiplayertools.ts [hunk 9]
```diff
@@ -774,15 +677,27 @@ namespace gdjs {
...
 
       // If we are connected to players, then the game can start.
       logger.info('Lobby game has started.');
+      // In case we're joining an existing lobby, read the saved messages to catch-up with the game state.
+      gdjs.multiplayerMessageManager.handleUpdateGameMessagesSaved(
+        runtimeScene
+      );
+      gdjs.multiplayerMessageManager.handleUpdateSceneMessagesSaved(
+        runtimeScene
+      );
```
Comment (4ian) [inline Extensions/Multiplayer/multiplayertools.ts:699]: I would hide `handleUpdateSceneMessagesSaved` and `handleUpdateGameMessagesSaved` from the interface of MultiplayerMessageManager, and have instead a single function `handleSavedUpdateMessages`.
No reason the caller is able to call this for the scene and not the game or vice versa.

File: Extensions/Multiplayer/messageManager.ts [hunk 10]
```diff
@@ -1768,7 +1862,17 @@ namespace gdjs {
...
+          if (gdjs.multiplayer.isReadyToSendOrReceiveGameUpdateMessages()) {
+            runtimeScene.getGame().updateFromNetworkSyncData(messageData);
+          } else {
+            // If the game is not ready to receive game update messages, we need to save the data for later use.
+            // This can happen when joining a game that is already running.
+            lastReceivedGameSyncDataUpdates.push(messageData);
+            if (lastReceivedGameSyncDataUpdates.length > 10) {
+              // We only keep the last 10 updates.
+              lastReceivedGameSyncDataUpdates.shift();
+            }
```
Comment (4ian) [inline Extensions/Multiplayer/messageManager.ts:1874]: This kind of code would almost mandate to be in a dedicated data structure so it's clearer for the reader that this is just a "either we update from what we received, or we store it for later":

```suggestion
            lastReceivedGameSyncDataUpdates.store(messageData);
```

File: Extensions/Multiplayer/multiplayertools.ts (hunk 6 above)
Comment (ClementPasteau) [reply Extensions/Multiplayer/multiplayertools.ts:78]: yup updating

File: Extensions/Multiplayer/messageManager.ts (hunk 8 above)
Comment (ClementPasteau) [reply Extensions/Multiplayer/messageManager.ts:2104]: it shouldn't crash and just return 0 as ping.

File: Extensions/Multiplayer/multiplayertools.ts (hunk 9 above)
Comment (ClementPasteau) [reply Extensions/Multiplayer/multiplayertools.ts:699]: done

File: Extensions/Multiplayer/messageManager.ts (hunk 10 above)
Comment (ClementPasteau) [reply Extensions/Multiplayer/messageManager.ts:1874]: done

**Pull request merged**